File: Knapsack.java
import java.util.Scanner;

public class Knapsack {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int numInstances = scanner.nextInt();

        for (int i = 0; i < numInstances; i++) {
            int numItems = scanner.nextInt();
            int capacity = scanner.nextInt();
            int[] weights = new int[numItems];
            int[] values = new int[numItems];

            for (int j = 0; j < numItems; j++) {
                weights[j] = scanner.nextInt();
                values[j] = scanner.nextInt();
            }

            int result = knapsack(numItems, capacity, weights, values);
            System.out.println(result);
        }
    }

    public static int knapsack(int numItems, int capacity, int[] weights, int[] values) {
        int[] dp = new int[capacity + 1];

        for (int i = 1; i <= numItems; i++) {
            for (int w = capacity; w >= weights[i - 1]; w--) {
                dp[w] = Math.max(dp[w], dp[w - weights[i - 1]] + values[i - 1]);
            }
        }

        return dp[capacity];
    }
}

File: Knapsack.py
def knapsack(num_instances, instances):
    for instance in instances:
        num_items, capacity, weights, values = instance
        dp = [0] * (capacity + 1)

        for i in range(1, num_items + 1):
            for w in range(capacity, weights[i - 1] - 1, -1):
                dp[w] = max(dp[w], dp[w - weights[i - 1]] + values[i - 1])

        print(dp[capacity])

if __name__ == "__main__":
    num_instances = int(input())
    instances = []

    for _ in range(num_instances):
        num_items, capacity = map(int, input().split())
        weights = []
        values = []
        for _ in range(num_items):
            weight, value = map(int, input().split())
            weights.append(weight)
            values.append(value)
        instances.append((num_items, capacity, weights, values))

    knapsack(num_instances, instances)